helix-core/src/main/java/org/apache/helix/controller/rebalancer/strategy/AbstractEvenDistributionRebalanceStrategy.java

```diff
@@ -82,7 +85,23 @@ public void init(String resourceName, final List<String> partitions,
   @Override
   public ZNRecord computePartitionAssignment(final List<String> allNodes,
       final List<String> liveNodes, final Map<String, Map<String, String>> currentMapping,
-      ResourceControllerDataProvider clusterData) throws HelixException {
+      ResourceControllerDataProvider clusterData) {
+    // only compute assignments for instances with available spaces
```

Comment:
Note that instance weight != available space. Just say filtering out all the instances with zero weight.